Held-Karp algorithm for TSP implementation

AlexandruValeanu opened the pull request:
I have implemented the dynamic programming algorithm for solving the general TSP (also known as the Held-Karp algorithm).

The algorithm can be seen on [wikipedia](https://en.wikipedia.org/wiki/Held%E2%80%93Karp_algorithm).

jkinable commented:
Before we can start the review process:
Please copy&paste a header into each of your new classes. Only change your name, nothing else! The date should read 2017-2017. A header can be found in any of the existing classes.

You state that the runtime complexity is O(2^|V|×|V|) and it takes O(2^|V|×|V|). However, shouldn't the runtime complexity be O(2^|V|×|V|^2)? Also, please use the math notation: $O(2^nn^2)$ and $O(2^nn)$. (See Math in Javadoc, https://github.com/jgrapht/jgrapht/wiki/Contributor-Guidelines)

AlexandruValeanu commented:
I have added the headers and updated the running time. I am using $|V|$ instead of $n$ so it would be consistent with the notation for `TwoMetricApproxTSP`.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+ *
+ * <p>
+ * See <a href="https://en.wikipedia.org/wiki/Held%E2%80%93Karp_algorithm">wikipedia</a> for more
+ * details about the dp algorithm.
```
Consider changing "dp" to dynamic programming.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+    public HeldKarpTSP() {
+    }
+
+    private void printState(int state, int n){
```
Consider deleting this, or moving this to the test package if really needed for debugging.

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+     * @throws IllegalArgumentException if the graph contains no vertices
+     */
+    @Override
+    public GraphPath<V, E> getTour(Graph<V, E> graph) {
```
Since state is kept in an integer, we need a check here to make sure that the user does not call this method with a very large graph.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+ * <p>
+ * This is an implementation of the Held-Karp algorithm which returns a minimum-cost Hamiltonian tour.
+ * The implementation requires the input graph to contain at least one vertex.
+ * The running time is $O(2^{|V|} \times |V|^2)$ and it takes $O(2^{|V|} \times |V|)$ extra memory.
```
Please rephrase the first sentence:
This is an implementation of the Held-Karp algorithm which returns an optimal, minimum-cost Hamiltonian tour.

Just to emphasize that this is an exact algorithm.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+
+            // map 'source' if no mapping exists
+            if (!mapToInt.containsKey(source)){
+                mapToInt.put(source, newNode);
```
can be substituted by mapToIn.put(source,mapToInt.size())

AlexandruValeanu replied:
I am not sure I understand what you mean. mapToIn.put(source,mapToInt.size()) used instead of mapToIn.put(source, newNode) is not an improvement (and you can't use .putIfAbsent).

jkinable replied:
The only difference is, that you don't need your `newNode` variable which gets incremented each time you add a node to the map (newNode is simply a counter which tracks the number of items in your map). This would make your code more concise (gets rid of 3 lines). Indeed not a 'performance' improvement.

AlexandruValeanu replied:
I have updated the code.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+         *   map each vertex to an integer (using a HashMap)
+         *   keep the reverse mapping  (using an ArrayList)
+         */
+        Map<V, Integer> mapToInt = new HashMap<>();
```
In several other implementations, the mapToInt and mapToV are called respectively:
vertexMap and indexList. Consider naming these the same.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+        List<E> edgeList = new ArrayList<>(n);
+
+        int lastNode = 0;
+        int lastState = 1 << lastNode;
```
since lastNode=0, 1 << 0 won't do anything.

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,210 @@
+import java.util.*;
+
+/**
+ * A dynamic programming algorithm for the TSP problem.
```
Do you have a reference as to which paper/article you used to create this particular implementation (please include)? Or did you solely use the wiki description?

AlexandruValeanu replied:
I only used the wiki description.

jkinable commented:
Nice implementation

jkinable commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -112,6 +103,10 @@ private double memo(int previousNode, int state, double[][] C, double[][] W){
         }
 
+        if (n > 31){
+            throw new IllegalArgumentException("Graph contains more than 31 vertices");
```
This is a really weird error message. Solving a TSP on a graph which has more than 31 vertices, is very common. Please make this error more descriptive, e.g.:
"The internal representation of the dynamic programming state space cannot represent graphs containing more than 31 vertices. The runtime complexity of this class (O(2^|V| x |V|^2)) makes it unsuitable for graphs with more than 31 vertices."

AlexandruValeanu replied:
Updated.

jkinable replied:
I don't think you need the quotes, (i.e. remove `\"`).

jkinable commented:
Looks fine to me. @d-michail if you don't have additional comments, could you merge this one?

d-michail commented on jgrapht-core/src/main/java/org/jgrapht/alg/tour/HeldKarpTSP.java:
```diff
@@ -0,0 +1,205 @@
+        }
+
+        if (n > 31){
+            throw new IllegalArgumentException("\"The internal representation of the dynamic programming state " +
```
Could you drop the quote here? Not really needed.

AlexandruValeanu replied:
Dropped.

d-michail commented:
Thanks for the nice contribution!
